### ISSUE TITLE
`test_segment` in `lyon_algorithms::hit_test` incorrectly assumes consecutive path segments

### ISSUE THREAD

**Issue by CircuitCoder** [nical/lyon#735]
In the current implementation of `test_segment`...:
https://github.com/nical/lyon/blob/master/crates/algorithms/src/hit_test.rs#L165
`prev_winding` only updates when the path segment pass the guard test at line 114. This causes that most path segment don't reset `prev_winding` to zero (or None).

A simple test case to showcase the incorrect result is to draw a rectangle around a point **two times**, as followed:
```rust
fn test_hit_test_consecutive() {
    use crate::math::point;
    use crate::path::Path;

    let mut builder = Path::builder();
    builder.begin(point(0.0, 0.0));
    builder.line_to(point(1.0, 0.0));
    builder.line_to(point(1.0, 1.0));
    builder.line_to(point(0.0, 1.0));
    builder.line_to(point(0.0, 0.0));
    builder.line_to(point(1.0, 0.0));
    builder.line_to(point(1.0, 1.0));
    builder.line_to(point(0.0, 1.0));
    builder.end(true);
    let path = builder.build();

    assert_eq!(path_winding_number_at_position(&point(0.5, 0.5), &path, 0.1), -2);
}
```

This test fails `-1 != -2`.

**Comment by CircuitCoder** [comment]
OK I've been trying to come up with some sort of fix to this problem, and run into the specific conterexample given in the comment in `test_segment`:
```
|
|
x-----x      p
      |
      |
```

So we really cannot reset `prev_winding` on every occasion. Maybe we can reset it only when the path segment pass through the other side of point p? (p.y \in [y1, y2] but p.x < path.sample(x))

**Comment by CircuitCoder** [comment]
So this is what I end up doing, which "works fine :tm:" for my use case.
```diff
diff --git a/crates/algorithms/src/hit_test.rs b/crates/algorithms/src/hit_test.rs
index 1c942f6..ae6ef33 100644
--- a/crates/algorithms/src/hit_test.rs
+++ b/crates/algorithms/src/hit_test.rs
@@ -113,12 +113,16 @@ fn test_segment(
     let y1 = segment.to.y;
     if f32::min(y0, y1) > point.y
         || f32::max(y0, y1) < point.y
-        || f32::min(segment.from.x, segment.to.x) > point.x
         || y0 == y1
     {
         return;
     }

+    if f32::min(segment.from.x, segment.to.x) > point.x {
+        *prev_winding = None;
+        return;
+    }
+
     let d = y1 - y0;

     let t = (point.y - y0) / d;
@@ -167,6 +171,8 @@ fn test_segment(
         }

         *prev_winding = Some(w);
+    } else {
+        *prev_winding = None;
     }
 }
```

**Comment by nical** [comment]
Hi, thanks for filing this and for the analysis. I had a look and I think that the solution is in the same vein. The main difference is that I think the `f32::max(y0, y1) < point.y` case should also reset `prev_winding`.

I'll put a fix up soon.

**Comment by nical** [comment]
The fix is published in `lyon_algorithms 0.17.7`.
